Let edit cursor follow focus in MIDI event list

leonardder opened the pull request:
It has always been a thorn in the eye for me that changing focus in the event list doesn't move the edit cursor, since you're basically moving focus in the event list to edit things, right? I added an option to allow that.
The new "Edit cursor follows focus in MIDI event list" option is disabled by default, as it modifies REAPER's default behavior in such a way that it could be potentially annoying/confusing for someone.

This pr is a draft until #596 is merged, as it depends on it and. Therefore a rebase is necessary after merging.

leonardder commented:
I didn't realize that this does not work on Mac. I now hid the option on Mac, which will add a gap in the dialog. I think this shouldn't be a big problem.

ScottChesworth commented:
Is there any chance of getting this to fly on Mac too? Ideally, I'd prefer to avoid features being platform-specific. The ability to jump platforms and keep workflows in tact is unique to OSARA in this space, I'd like to preserve that wherever possible.

leonardder commented:
We do not support note preview on event list focus either. I'm afraid 
this is simply not possible unless Cockos gives us a way to be modified 
about focus changes in the REAPER interface.

ScottChesworth commented:
> We do not support note preview on event list focus either.
> 
Ah, somehow I didn't know that. Well if we're already down the rabbit hole and you've already done the work, then I retract my gripe. :)

Lo-lo78 commented:
👍👍👍

jcsteh commented:
Does this force the edit cursor to go to the start of the MIDI item when you open the event list? If I remember correctly, the first event gets focused in that case. That wouldn't be ideal.

Even if it doesn't, if the user presses down arrow without first pressing control+f to focus the nearest event to the cursor, they'll get thrown to the start of the item then. That said, I don't know that there's a good solution to this. I wondered about auto focusing the nearest event when opening the event list, but I think that would be difficult to manage without getting into nasty conflict situations.

leonardder commented:
> Does this force the edit cursor to go to the start of the MIDI item when you open the event list? If I remember correctly, the first event gets focused in that case. That wouldn't be ideal.

Ah, that's a good catch ineed.

> I wondered about auto focusing the nearest event when opening the event list, but I think that would be difficult to manage without getting into nasty conflict situations.

Could you elaborate on what conflict situations might arise?

jcsteh commented:
I think my thoughts about conflicts related to the fact that if we incorrectly focus the nearest event for a focus other than the initial focus, we could end up in a situation where we focus the nearest event, that triggers a focus event, we respond to that by focusing the nearest event. However, I now realise it'd stop there rather than getting into a loop because the focus wouldn't actually change in the final part. Still, that would be pretty nasty regardless.

leonardder commented:
As far as I can see, on initial focus, no actual item in the list gets focus, so that case should be easy to detect.

Lo-lo78 commented:
Hi Leonard, I just tried your latest build and with midi files with many 
events in one item there is a latency. I think it depends on the number 
of events in a midi item; if I create smaller items there is no latency.

Greetings!

jcsteh commented:
Are you seeing latency when using the cursor keys or only when first opening the event list? I can understand latency when opening (since focusing the nearest event has to search many events), but I'd be surprised by latency when using the cursor keys.

jcsteh commented:
If the latency is caused by focusing the nearest event, we should be able to improve that using a binary search. I implemented binary search for the piano roll in #391. I never landed it because no one ever confirmed that it helped them in real usage. However, with the framework I implemented there, it should be possible to create a template to do this for the event list.

Lo-lo78 commented:
Yes, there is latency when I open the event list, when it ends it can be 
put into play, which causes an interruption for the play as well.

ScottChesworth commented:
@Lo-lo78, can you link to a MIDI file we can use for consistency during testing? Many events will mean different things to different users.

Lo-lo78 commented:
Here is the link:

https://drive.google.com/uc?id=11sVbS5wxNbzJqKmZqIClh17UN_GVDBs1&export=download

ScottChesworth commented:
Thanks. Now, can you test to see if the event list is still responding slowly loading this data with OSARA pre621 from the snapshots page?
Here, the event list is laggy irrespective of whether I'm using pre621 or the most recent AppVeyor build in this thread. On my computer, @leonardder's contribution doesn't make the situation noticeably better or worse. Are you seeing the same thing?

leonardder commented:
I did a force push because it looks like appveyor is hanging once again.

leonardder commented:
> If the latency is caused by focusing the nearest event, we should be able to improve that using a binary search. I implemented binary search for the piano roll in #391. I never landed it because no one ever confirmed that it helped them in real usage. However, with the framework I implemented there, it should be possible to create a template to do this for the event list.

I think the binary search can also help with fixing #434 right? Anything against merging that and let it prove itself?

ScottChesworth commented:
> Anything against merging that and let it prove itself?
> 
Just tried to download the windows build in #393 to test whether it improves latency with the MIDI file in this thread. It's been more than a month, so AppVeyor doesn't have the build anymore. How do we rebuild that?
BTW, sorry I missed it first time round @jcsteh.

leonardder commented:
Think you meant #391? I merged latest master into it. However, note that binary search is only implemented for notes/chords and not for events.

Lo-lo78 commented:
Hi, now with the screen reader there is no latency but if I switch from 
the piano roll to the event list the play is slowed down the first time 
always with large items. With smaller items there is no problem.

jcsteh commented:
> I think the binary search can also help with fixing #434 right? Anything against merging that and let it prove itself?

As you realised, the binary search isn't currently implemented for CCs. However, the framework should make implementing that a lot easier. I've merged it now because it seems to help @Lo-lo78 and so we can base other improvements on that framework.

leonardder commented:
@jcsteh is there anything holding this back currently?

jcsteh commented:
As I understand it, there are still performance concerns as reported in https://github.com/jcsteh/osara/pull/599#issuecomment-970673460. Fixing this would require chanhing to a binary search here. If you did that already, I didn't see any commit notifications here...

leonardder commented:
@Lo-lo78 could you please try the try build that will be posted shortly after this message to see how performance is going compared to the build of two days ago?

leonardder commented:
Ah, it failed on mac. Never mind, here is the windows build to test with: https://ci.appveyor.com/api/buildjobs/s0gusns4l1v4fjsf/artifacts/installer%2Fosara_pr599-758%2C3508b6cd.exe

Lo-lo78 commented:
@leonardder 
Great, the piano roll motor is very snappy and the first time it is perfect it synchronizes.
However, I do not think it is intended, without exiting the midi editor, if I reopen the piano roll and I move and re-enter the event list, I go back to the previous point.
For instance:
Piano Roll: event 307
Event List: event 307
Piano Roll: event 106
Event list: event 307.

If I do it for the second time from the piano roll and I go to the event list it doesn't sync.
I have to move the event list cursor to update the position otherwise it returns to event 307.
;))
Am I wrong something?

leonardder commented:
Could you provide the midi file to reproduce?

Lo-lo78 commented:
Hi, try this midi file.
[Diana Est - Tenax3.zip](https://github.com/jcsteh/osara/files/7906022/Diana.Est.-.Tenax3.zip)

leonardder commented:
I think I covered all edge cases now and indeed, the binary search is really fast and snappy!

@Lo-lo78 I think I fixed your case as well, would be great if you could verify. I also made sure that no movement takes place when the current focus is at the cursor position. The binary search iterator together with equal_range makes thas pretty elegant if I may say so.

Lo-lo78 commented:
Very good!

Congratulations, you have improved the situation a lot!

My opinion is that you have solved, look forward to the opinions of others.

For me it's perfect!

👍👍👍

Message ID: ***@***.***>

jcsteh commented on src/reaper_osara.rc:
```diff
@@ -63,8 +63,11 @@ BEGIN
 	CONTROL "Report changes made via &control surfaces", ID_CONFIG_REPORT_SURFACE_CHANGES, "Button", BS_AUTOCHECKBOX | WS_TABSTOP, 10, 166, 155, 14
-	DEFPUSHBUTTON "OK", IDOK, 10, 186, 30, 14
-	PUSHBUTTON "Cancel", IDCANCEL, 137, 186, 40, 14
+#ifdef _WIN32
```
Does the PHP script which generates SWELL resources respect ifdefs? In other words, do we know for certain that this actually excludes the control on Mac?

leonardder replied:
AH that's actually pretty unlikely looking at the script. I wonder how to fix this.

jcsteh replied:
I think you'll need to do this at runtime, after the dialog is created but before it is shown. If Mac, get the HWND for the checkbox and hide it. You should be able to use ShowWindow with the right arguments to hide it, though I don't recall the specifics. You could test that on Windows easily enough. Failing that, you could just disable it with EnableWindow(hwnd, false) which I know for a fact works because I do that elsewhere.

jcsteh reviewed (commented):
This is looking great.

leonardder commented:
I first started with EnableWindow(.., FALSE), then realizing that it still shows the window text. ShowWindow works much better indeed.

jcsteh reviewed (approved):
Thanks so much for working on this.

jcsteh commented:
> The binary search iterator together with equal_range makes thas pretty elegant if I may say so.

Agreed. When I wrote the original version of the binary search code, I spent a fair bit of time on it and it then languished for ages without feedback. I'm glad it has proven to be useful and grateful that you've taken it to the next level. :)

leonardder commented:
@RDMurray is this mac or Windows?

leonardder commented:
Ugh you are correct. Pretty sure this didn't work before. Anyhow, I guess we can revert this.

Lo-lo78 commented:
Bye!
Yes. It works even without OSARA.
Too bad that Reaper has not completely resolved the latency between the transition from the piano roll to the event list. With large items that have large lists, Reaper has some problems.
the Actions List is Reaper's only quick list. When you add a plugin there is latency. Maybe they'll fix these big lists too. ;))
👍
